client.go

```diff
@@ -150,37 +166,103 @@ func (cfg *Config) hostKey() ssh.HostKeyCallback {
 	return ssh.InsecureIgnoreHostKey()
 }
 
+// saves SSH client so it can be closed later
+func (s *SessionInfo) saveClient(sshClient *ssh.Client) {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	s.openClients = append(s.openClients, sshClient)
+}
+
+// saves SSH connection so it can be closed later
+func (s *SessionInfo) saveConn(conn net.Conn) {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	s.openConns = append(s.openConns, conn)
+}
+
+func (s *SessionInfo) CloseAll() {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	for _, cl := range s.openClients {
+		cl.Close()
+	}
+	for _, con := range s.openConns {
+		con.Close()
+	}
+	s.openClients = nil
+	s.openConns = nil
+	if s.openSession != nil {
+		s.openSession.Close()
+	}
+	s.openSession = nil
+
+}
+
+// Copy copies the NativeClient with empty SessionInfo
+func (c *NativeClient) Copy() *NativeClient {
+	// copy the existing host details
+	hds := make([]HostDetail, len(c.HostDetails))
+	copy(hds, c.HostDetails)
+
+	var sessionInfo SessionInfo
+	var copyClient = NativeClient{
+		HostDetails:         hds,
+		ClientVersion:       c.ClientVersion,
+		DefaultClientConfig: c.DefaultClientConfig,
+		SessionInfo:         &sessionInfo,
+	}
+	return &copyClient
+}
+
+// AddHopWithConfig adds a new host to the end of the list and returns a new client using the provided config
+// The original client is unchanged
+func (c *NativeClient) AddHopWithConfig(host string, port int, config *ssh.ClientConfig) (interface{}, error) {
```

Comment:
This (and AddHop) should return Client (the interface defined on line 72) instead of interface{}. This matches what NewNativeClient returns.

Reply:
Ok, I see this conflicts with the definition of AddHop in PlatformClient. Hm, let me think about this.

Reply:
aside from making it an interface, the other choice I think would be to have a definition used commonly between golang-ssh and the other repos

Reply:
Yeah, that would be the typical way to do this. Have the PlatformClient interface imported by both localClient and sshClient. You'd have to do it that way in Java, "public Class NativeClient implements PlatformClient". I think for sure we'd want to do that for platform.local because it's in the same repo.
The other way we could do it (if we don't want to import PlatformClient into golang-ssh) is to wrap golang-ssh client inside a new edgeproto object, like SshClient. You can override the AddHop method while inheriting the other methods (tried an example in go playground and seems to compile ok). So in golang-ssh:
```
type Client interface {
      ...
      AddHop(...) (Client, error)
}
```

In edgecloud:
```
// SshClient implements PlatformClient
type SshClient struct {
        ssh.Client // inherit golang-ssh Client funcs
}

func (s *SshClient) NewClient(...) (*SshClient, error) {
           gossh, err := ssh.NewNativeClient(...)
           if err != nil {
                      return nil, err
           }
           return &SshClient{Client: gossh}, nil
}

// override ssh.Client AddHop() func
func (s *SshClient) AddHop(...) (PlatformClient, error) {
         // call golang-ssh AddHop
         gossh, err := s.Client.AddHop(...)
         if err != nil {
                    return err
         }
         return &SshClient{Client: gossh}, nil
}
```
But really the "proper" way to do this is what you mentioned, so that might be the better way to do it. That would just mean here in golang-ssh, AddHop would return edgecloud.PlatformClient.